src/coreclr/jit/lclvars.cpp

```diff
@@ -1790,6 +1788,62 @@ bool Compiler::StructPromotionHelper::CanPromoteStructType(CORINFO_CLASS_HANDLE
     return true;
 }
 
+//--------------------------------------------------------------------------------------------
+// CanConstructAndPromoteField - checks if we can construct field types without asking about them directly.
+//
+// Arguments:
+//   structPromotionInfo - struct promotion candidate information.
+//
+// Return value:
+//   true if we can figure out the fields from available knowledge.
+//
+// Notes:
+//   This is needed for AOT R2R compilation when we can't cross compilation bubble borders
+//   so we should not ask about fields that are not directly referenced. If we do VM will have
+//   to emit a type check for this field type but it does not have enough information about it.
+//   As a workaround for perfomance critical corner case: struct with 1 gcref, we try to construct
+//   the field information from indirect observations.
+//
+bool Compiler::StructPromotionHelper::CanConstructAndPromoteField(lvaStructPromotionInfo* structPromotionInfo)
+{
+    const CORINFO_CLASS_HANDLE typeHnd    = structPromotionInfo->typeHnd;
+    const COMP_HANDLE          compHandle = compiler->info.compCompHnd;
+    const DWORD                typeFlags  = compHandle->getClassAttribs(typeHnd);
```

Comment:
Strictly speaking <code>getClassAttribs</code> should probably also trigger emission of the type layout check - if JIT queried this information and made some codegen decisions based on it, the generated code should be also invalidate when the class attributes change. I believe it's not happening today and it may be the case that all "important" places asking for <code>getClassAttribs</code> ultimately also need to query <code>getClassSize</code> but it's still pretty fragile. Adding @davidwrighton to hear his thoughts on the subject.